### ISSUE TITLE
threads->tid incorrect in DESTROY

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5716]
Migrated from [rt.perl.org#10046](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=10046) (status was 'resolved')

Searchable as RT10046$

**Comment by p5pRT** [comment]
### From @lizmat

#### Created by @lizmat



It seems that threads\->tid is reported incorrectly inside a DESTROY\.
The tid reported is the tid in which the object was originally created\.

This basically forces the object to keep a flag itself whether any
shared parts of the object are already destroyed or not\.  And may
have implications for modules such as DBI being used in an environment
where there are threads active\.

Set the severity to "medium"\, but this may have bigger implications
which might warrant an increase in severity\.  ;\-\(

== tidwrong =====================================================
use threads;

my $async = async \{
my $foo = Foo\->new;
my $thread = threads\->new\( sub \{print "in thread "\.threads\->tid\."\\n"\} \);
print "waiting for thread to finish\\n";
$thread\->join;
print "thread now finished\\n";
\};

$async\->join;

package Foo;
sub new \{ bless \{\}\,shift \};
sub DESTROY \{ print "DESTROYing in thread "\.threads\->tid\."\\n"\}
=================================================================

&nbsp;  $ perl \-w tidwrong
&nbsp;  waiting for thread to finish
&nbsp;  in thread 2
&nbsp;  DESTROYing in thread 1
&nbsp;  thread now finished
&nbsp;  DESTROYing in thread 1

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=medium

Site configuration information for perl v5.8.0:

Configured by liz at Wed Jul 10 18:32:18 CEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0 patch 17462) 
configuration:
   Platform:
     osname=linux, osvers=2.4.18, archname=i686-linux-thread-multi
     uname='linux echt.dijkmat.nl 2.4.18 #8 smp mon mar 25 22:28:36 cet 
2002 i686 unknown '
     config_args='-de -Dusethreads'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing 
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 
-I/usr/include/gdbm',
     optimize='-O2',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing 
-I/usr/local/include -I/usr/include/gdbm'
     ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-98)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib
     libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
     perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
     libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.2.4'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
     cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
     DEVEL17462


@INC for perl v5.8.0:
     /usr/local/lib/perl5/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/5.8.0
     /usr/local/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/site_perl/5.8.0
     /usr/local/lib/perl5/site_perl/5.7.3
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.8.0:
     HOME=/home/liz
     LANG=en_US
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/liz/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash




```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hah\, this sounds interesting\,

I think you will find it reports the tid of the thread that does the 
join :\-\)\, I don't know offhand how to fix it\.

Arthur

On fredag\, juli 12\, 2002\, at 05&#8203;:31 \, Jarkko Hietaniemi wrote&#8203;:

> \*sigh\*  Why do people have to do this\, find more and more bugs
> just before a release\.\.\.?
>
> \-\-\-\-\- Forwarded message from Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> \-\-\-\-\-
>
> Subject&#8203;: \[ID 20020712\.002\] threads\->tid incorrect in DESTROY
> From&#8203;: Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl>
> Date&#8203;: Fri\, 12 Jul 2002 17&#8203;:14&#8203;:27 \+0200
> Message\-Id&#8203;: \<4\.2\.0\.58\.20020712171338\.02556380@&#8203;mickey\.dijkmat\.nl>
> To&#8203;: perl5\-porters@&#8203;perl\.org
>
> This is a bug report for perl from liz@&#8203;dijkmat\.nl\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> It seems that threads\->tid is reported incorrectly inside a DESTROY\.
> The tid reported is the tid in which the object was originally created\.
>
> This basically forces the object to keep a flag itself whether any
> shared parts of the object are already destroyed or not\.  And may
> have implications for modules such as DBI being used in an environment
> where there are threads active\.
>
> Set the severity to "medium"\, but this may have bigger implications
> which might warrant an increase in severity\.  ;\-\(
>
> == tidwrong =====================================================
> use threads;
>
> my $async = async \{
> my $foo = Foo\->new;
> my $thread = threads\->new\( sub \{print "in thread "\.threads\->tid\."\\n"\} \);
> print "waiting for thread to finish\\n";
> $thread\->join;
> print "thread now finished\\n";
> \};
>
> $async\->join;
>
> package Foo;
> sub new \{ bless \{\}\,shift \};
> sub DESTROY \{ print "DESTROYing in thread "\.threads\->tid\."\\n"\}
> =================================================================
>
>   $ perl \-w tidwrong
>   waiting for thread to finish
>   in thread 2
>   DESTROYing in thread 1
>   thread now finished
>   DESTROYing in thread 1
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>      category=core
>      severity=medium
> \-\-\-
> Site configuration information for perl v5\.8\.0&#8203;:
>
> Configured by liz at Wed Jul 10 18&#8203;:32&#8203;:18 CEST 2002\.
>
> Summary of my perl5 \(revision 5\.0 version 8 subversion 0 patch 17462\)
> configuration&#8203;:
>    Platform&#8203;:
>      osname=linux\, osvers=2\.4\.18\, archname=i686\-linux\-thread\-multi
>      uname='linux echt\.dijkmat\.nl 2\.4\.18 \#8 smp mon mar 25 22&#8203;:28&#8203;:36 cet
> 2002 i686 unknown '
>      config\_args='\-de \-Dusethreads'
>      hint=recommended\, useposix=true\, d\_sigaction=define
>      usethreads=define use5005threads=undef useithreads=define
> usemultiplicity=define
>      useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>      use64bitint=undef use64bitall=undef uselongdouble=undef
>      usemymalloc=n\, bincompat5005=undef
>    Compiler&#8203;:
>      cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing
> \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64
> \-I/usr/include/gdbm'\,
>      optimize='\-O2'\,
>      cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing
> \-I/usr/local/include \-I/usr/include/gdbm'
>      ccversion=''\, gccversion='2\.96 20000731 \(Red Hat Linux 7\.1 
> 2\.96\-98\)'\,
> gccosandvers=''
>      intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>      d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>      ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
> lseeksize=8
>      alignbytes=4\, prototype=define
>    Linker and Libraries&#8203;:
>      ld='cc'\, ldflags =' \-L/usr/local/lib'
>      libpth=/usr/local/lib /lib /usr/lib
>      libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>      perllibs=\-lnsl \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>      libc=/lib/libc\-2\.2\.4\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>      gnulibc\_version='2\.2\.4'
>    Dynamic Linking&#8203;:
>      dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, 
> ccdlflags='\-rdynamic'
>      cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
>
> Locally applied patches&#8203;:
>      DEVEL17462
>
> \-\-\-
> @&#8203;INC for perl v5\.8\.0&#8203;:
>      /usr/local/lib/perl5/5\.8\.0/i686\-linux\-thread\-multi
>      /usr/local/lib/perl5/5\.8\.0
>      /usr/local/lib/perl5/site\_perl/5\.8\.0/i686\-linux\-thread\-multi
>      /usr/local/lib/perl5/site\_perl/5\.8\.0
>      /usr/local/lib/perl5/site\_perl/5\.7\.3
>      /usr/local/lib/perl5/site\_perl
>      \.
>
> \-\-\-
> Environment for perl v5\.8\.0&#8203;:
>      HOME=/home/liz
>      LANG=en\_US
>      LANGUAGE \(unset\)
>      LD\_LIBRARY\_PATH \(unset\)
>      LOGDIR \(unset\)
>      PATH=/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/home/liz/bin
>      PERL\_BADLANG \(unset\)
>      SHELL=/bin/bash
>
> \-\-\-\-\- End forwarded message \-\-\-\-\-
>
> \-\-
> $jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
>         \# There is this special biologist word we use for 'stable'\.
>         \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From @lizmat

At 05&#8203;:14 PM 7/12/02 \+0200\, you wrote&#8203;:
>This is a bug report for perl from liz@&#8203;dijkmat\.nl\,
>generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>It seems that threads\->tid is reported incorrectly inside a DESTROY\.
>The tid reported is the tid in which the object was originally created\.

The same problem seems to exist with the CLONE method\.



Liz

**Comment by p5pRT** [comment]
### From @lizmat

At 04&#8203;:38 PM 7/21/02 \+0200\, Elizabeth Mattijsen wrote&#8203;:
>At 05&#8203;:14 PM 7/12/02 \+0200\, you wrote&#8203;:
>>It seems that threads\->tid is reported incorrectly inside a DESTROY\.
>>The tid reported is the tid in which the object was originally created\.
>The same problem seems to exist with the CLONE method\.

With my first attempts at XS\, I found out that the problem is actually even 
deeper\.  The entire context is still in the originating thread\, judging 
from the value of the C getpid\(\) function\.

perlmod\.pod states&#8203;:

"CLONE will be executed once for every package that has it defined \(or 
inherits it\)\. It will be called in the context of the new thread\, so all 
modifications are made in the new area\."

This may be true at some Perl level\, it definitely does not seem to be true 
at the C level \(I have no other explanation for getpid\(\) showing the pid of 
the originating thread's process under Linux\)\.


Liz

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On måndag\, september 2\, 2002\, at 10&#8203;:07 \, Elizabeth Mattijsen wrote&#8203;:

>
> With my first attempts at XS\, I found out that the problem is actually 
> even deeper\.  The entire context is still in the originating thread\, 
> judging from the value of the C getpid\(\) function\.
>

Uhm\, please do not confuse the perl context with the underlying thread 
context\, in theory we are free to migrate perl threads between 
underlying OS threads \(mod\_perl does that\)

> perlmod\.pod states&#8203;:
>
> "CLONE will be executed once for every package that has it defined \(or 
> inherits it\)\. It will be called in the context of the new thread\, so 
> all modifications are made in the new area\."
>
> This may be true at some Perl level\, it definitely does not seem to be 
> true at the C level \(I have no other explanation for getpid\(\) showing 
> the pid of the originating thread's process under Linux\)\.
>
>

The cloning is done in a new context but in the old os thread\.

Arthur

**Comment by p5pRT** [comment]
### From @lizmat

At 11&#8203;:53 AM 9/2/02 \+0200\, Arthur Bergman wrote&#8203;:
>>This may be true at some Perl level\, it definitely does not seem to be 
>>true at the C level \(I have no other explanation for getpid\(\) showing the 
>>pid of the originating thread's process under Linux\)\.
>The cloning is done in a new context but in the old os thread\.

I assume that's the same for a CLONE \{\} that I write in XS?


Liz

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On måndag\, september 2\, 2002\, at 12&#8203;:04 \, Elizabeth Mattijsen wrote&#8203;:

>
> I assume that's the same for a CLONE \{\} that I write in XS?
>
>

Indeed\.

Arthur

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:12 PM 9/2/02 \+0200\, Arthur Bergman wrote&#8203;:
>>I assume that's the same for a CLONE \{\} that I write in XS?
>Indeed\.

Would it be possible to get at the new thread id and the getpid\(\) of the 
new thread in XS from within a CLONE subroutine?


Liz

**Comment by p5pRT** [comment]
### From goldbb2@earthlink.net

Elizabeth Mattijsen wrote&#8203;:
> 
> At 01&#8203;:12 PM 9/2/02 \+0200\, Arthur Bergman wrote&#8203;:
> >>I assume that's the same for a CLONE \{\} that I write in XS?
> >Indeed\.
> 
> Would it be possible to get at the new thread id and the getpid\(\) of
> the new thread in XS from within a CLONE subroutine?

Within the CLONE method\, the "current thread id" should \*be\* the "new
thread id"\.  Regardless of whether or not doing it in XS\.

And the new OS\-thread may not yet have been created\.

\-\- 
tr/\`4/ /d\, print "@&#8203;\{\[map \-\-$| ? ucfirst lc : lc\, split\]\}\,\\n" for
pack 'u'\, pack 'H\*'\, 'ab5cf4021bafd28972030972b00a218eb9720000';

**Comment by p5pRT** [comment]
### From arthur@contiller.se


On måndag\, september 2\, 2002\, at 11&#8203;:55 \, Elizabeth Mattijsen wrote&#8203;:

>
> At 01&#8203;:12 PM 9/2/02 \+0200\, Arthur Bergman wrote&#8203;:
>>> I assume that's the same for a CLONE \{\} that I write in XS?
>> Indeed\.
>
> Would it be possible to get at the new thread id and the getpid\(\) of 
> the new thread in XS from within a CLONE subroutine?
>
>

Not the getpid since the process thread is not created yet\.

Arthur

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> writes&#8203;:
>At 01&#8203;:12 PM 9/2/02 \+0200\, Arthur Bergman wrote&#8203;:
>>>I assume that's the same for a CLONE \{\} that I write in XS?
>>Indeed\.
>
>Would it be possible to get at the new thread id and the getpid\(\) of the 
>new thread in XS from within a CLONE subroutine?

Not at present\. The new thread has not been created at point CLONE is called\.
We call CLONE to create data area for new thread\. As soon a new thread 
is created it starts running and accessing data so it needs to have 
somewhere to play\. 

I can think of two alternatives&#8203;:
A\. Suspend the original thread \(to keep it out of to\-be\-cloned data\)
&nbsp;   and do CLONE in chiled thread\, then awaken parent thread\.
&nbsp;   \(A bit like doing a vfork rather than fork I guess\.\)

B\. Invent CLONED \(say\) which is run in child thread before calling 
&nbsp;   the sub\. Maybe we could generalize this to have 
&nbsp;   PRECLONE \{ \} and POSTCLONE \{ \} lists akin 
&nbsp;   BEGIN \{\} CHECK \{\} INIT \{\} END \{\}



>
>
>Liz
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @lizmat

At 06&#8203;:25 PM 9/2/02 \-0400\, Benjamin Goldberg wrote&#8203;:
> > Would it be possible to get at the new thread id and the getpid\(\) of
> > the new thread in XS from within a CLONE subroutine?
>Within the CLONE method\, the "current thread id" should \*be\* the "new
>thread id"\.  Regardless of whether or not doing it in XS\.

Well\, I agree\.  But that is \_not\_ the case right now\.


>And the new OS\-thread may not yet have been created\.

Well\, after some more thinking\, I created a work around that seems to 
work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the 
thread before the actual thread subroutine is started\.

Why wasn't CLONE implemented that way?


Liz

**Comment by p5pRT** [comment]
### From @rgarcia

Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> wrote&#8203;:
> 
> Well\, after some more thinking\, I created a work around that seems to 
> work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the 
> thread before the actual thread subroutine is started\.

Do you mean that you subclass threads ? Is it possible to subclass
threads ? \(BTW thread objects are done this way in Java\.\)

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:24 PM 9/3/02 \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> > Well\, after some more thinking\, I created a work around that seems to
> > work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the
> > thread before the actual thread subroutine is started\.
>Do you mean that you subclass threads ? Is it possible to subclass
>threads ? \(BTW thread objects are done this way in Java\.\)

Well\, you \_could\_ subclass threads\.pm\, why not?

But what I was thinking\, was more along the lines of&#8203;:

&nbsp;   use threads;
&nbsp;   my $new = \\&threads&#8203;::new;
&nbsp;   \*threads&#8203;::new = sub \{
&nbsp;     my $class = shift;
&nbsp;     my $sub = shift;
&nbsp;     \* do whatever I want to do inside the thread here
&nbsp;     goto &$sub;
&nbsp;   \};

With this trick I can&#8203;:

&nbsp;  \- reliably get signals to automatically register in a thread
&nbsp;    for Thread&#8203;::Status\, which will allow you to obtain a live
&nbsp;    report of where each thread is running in the code by just
&nbsp;    specifying \-MThread&#8203;::Status in the command line\.

&nbsp;  \- allow me to return from anywhere in a thread wthout getting
&nbsp;    a message to STDERR \(anyone objects to stealing exit\(\) for
&nbsp;    this?\)\, which I'll probably call Thread&#8203;::Exit\.

&nbsp;  \- lot's of other nice stuff involving events\.


Liz

**Comment by p5pRT** [comment]
### From @tux

On Wed 04 Sep 2002 01&#8203;:32\, Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> wrote&#8203;:
> At 01&#8203;:24 PM 9/3/02 \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> > > Well\, after some more thinking\, I created a work around that seems to
> > > work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the
> > > thread before the actual thread subroutine is started\.
> >Do you mean that you subclass threads ? Is it possible to subclass
> >threads ? \(BTW thread objects are done this way in Java\.\)
> 
> Well\, you \_could\_ subclass threads\.pm\, why not?
> 
> But what I was thinking\, was more along the lines of&#8203;:
> 
>    use threads;
>    my $new = \\&threads&#8203;::new;
>    \*threads&#8203;::new = sub \{
>      my $class = shift;
>      my $sub = shift;
>      \* do whatever I want to do inside the thread here
>      goto &$sub;

And where do you use $new now?

>    \};
> 
> With this trick I can&#8203;:
> 
>   \- reliably get signals to automatically register in a thread
>     for Thread&#8203;::Status\, which will allow you to obtain a live
>     report of where each thread is running in the code by just
>     specifying \-MThread&#8203;::Status in the command line\.
> 
>   \- allow me to return from anywhere in a thread wthout getting
>     a message to STDERR \(anyone objects to stealing exit\(\) for
>     this?\)\, which I'll probably call Thread&#8203;::Exit\.
> 
>   \- lot's of other nice stuff involving events\.
> 
> 
> Liz

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.0 & 633 on HP\-UX 10\.20 & 11\.00\, AIX 4\.2\, AIX 4\.3\,
&nbsp;  WinNT 4\, Win2K pro & WinCE 2\.11\.  Smoking perl CORE&#8203;: smokers@&#8203;perl\.org
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From @lizmat

At 08&#8203;:45 AM 9/4/02 \+0200\, H\.Merijn Brand wrote&#8203;:
> > But what I was thinking\, was more along the lines of&#8203;:
>And where do you use $new now?

Oops\.  I should have known better than to write this down from memory after 
a 2 hour drive in the middle of the night\.  Now that my development machine 
is up\, I can cut and past what I thought of earlier yesterday&#8203;:

&nbsp;  my $new = \\&threads&#8203;::new;
&nbsp;  \*threads&#8203;::new = sub \{
&nbsp;      my $class = shift;
&nbsp;      my $sub = shift;
&nbsp;      $new\->\( $class\,sub \{
&nbsp;          \* do what I want to do inside the thread here
&nbsp;          goto &$sub;
&nbsp;      \}\,@&#8203;\_ \);


Liz

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> writes&#8203;:
>At 06&#8203;:25 PM 9/2/02 \-0400\, Benjamin Goldberg wrote&#8203;:
>> > Would it be possible to get at the new thread id and the getpid\(\) of
>> > the new thread in XS from within a CLONE subroutine?
>>Within the CLONE method\, the "current thread id" should \*be\* the "new
>>thread id"\.  Regardless of whether or not doing it in XS\.
>
>Well\, I agree\.  But that is \_not\_ the case right now\.
>
>
>>And the new OS\-thread may not yet have been created\.
>
>Well\, after some more thinking\, I created a work around that seems to 
>work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the 
>thread before the actual thread subroutine is started\.

It might make sense for threads&#8203;::new to always call threads&#8203;::something
before calling user's sub\.

>
>Why wasn't CLONE implemented that way?

As I tried to explain before \- if child thread is to get a true "clone"
at instant it was created \(and indeed to avoid races on these un\-shared things\)
the parent thread must be suspended and not modifying things while it happens\, 
and then re\-awoken\. 

The current CLONE fakes this by having the parent thread do the clone
\(so it isn't doing anything else\)\.

\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @lizmat

At 04&#8203;:13 PM 9/5/02 \+0100\, Nick Ing\-Simmons wrote&#8203;:
> >Well\, after some more thinking\, I created a work around that seems to
> >work\.  By hijacking threads&#8203;::new\, you can run a subroutine inside the
> >thread before the actual thread subroutine is started\.
>It might make sense for threads&#8203;::new to always call threads&#8203;::something
>before calling user's sub\.

Hmmm\.\.\. I could add a begin\(\) method to Thread&#8203;::Exit\, analogous to the 
end\(\) class method\.  Writing the documentation would be more work\.


> >Why wasn't CLONE implemented that way?
>The current CLONE fakes this by having the parent thread do the clone
>\(so it isn't doing anything else\)\.

Now that makes sense to me\.


Liz

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Hi\, This is now solved\.

We do this by faking what thread we are running in during destroy and
clone operations so the tid is correctly reported\.

However you will not get a correct underlying OS tid\, nor a correct pid
in linux\, however one should not depend on the underlying OS tid\,
neither should one look at the pid since we are free to migrate perl
threads across OS level threads\.

The patch that fixes this is \#18417

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
